src/neptune/core/operation_processors/async_operation_processor.py

```diff
@@ -0,0 +1,430 @@
+        )
+
+        self._queue_observer = QueueObserver(
+            disk_queue=self._processing_resources.disk_queue,
+            consumer=self._consumer,
+            should_print_logs=self._should_print_logs,
+            stop_queue_max_time_no_connection_seconds=self.STOP_QUEUE_MAX_TIME_NO_CONNECTION_SECONDS,
+        )
+
+    @property
+    def resources(self) -> Tuple[Resource, ...]:
+        return self._processing_resources.resources
+
+    @property
+    def data_path(self) -> Path:
+        return self._processing_resources.data_path
+
+    @property
+    def processing_resources(self) -> "ProcessingResources":
+        return self._processing_resources
+
+    @ensure_disk_not_overutilize
+    def enqueue_operation(self, op: Operation, *, wait: bool) -> None:
+        if not self._accepts_operations:
+            warn_once("Not accepting operations", exception=NeptuneWarning)
+            return
+
+        self._last_version = self.processing_resources.disk_queue.put(op)
+
+        if _queue_has_enough_space(self.processing_resources.disk_queue.size(), self._processing_resources.batch_size):
+            self._consumer.wake_up()
+        if wait:
+            self.wait()
+
+    def start(self) -> None:
+        self._consumer.start()
+
+    def pause(self) -> None:
+        self._consumer.pause()
+        self.flush()
+
+    def resume(self) -> None:
+        self._consumer.resume()
+
+    def wait(self) -> None:
+        self.flush()
+        waiting_for_version = self._last_version
+        self._consumer.wake_up()
+
+        # Probably reentering lock just for sure
+        with self._processing_resources.waiting_cond:
+            self._processing_resources.waiting_cond.wait_for(
+                lambda: self._processing_resources.consumed_version >= waiting_for_version
+                or not self._consumer.is_running()
+            )
+        if not self._consumer.is_running():
+            raise NeptuneSynchronizationAlreadyStoppedException()
+
+    def stop(
+        self,
+        seconds: Optional[float] = None,
+        processor_stop_signal_queue: Optional["Queue[ProcessorStopSignal]"] = None,
+    ) -> None:
+        ts = time()
+        self.flush()
+        if self._consumer.is_running():
+            self._consumer.disable_sleep()
+            self._consumer.wake_up()
+            self._queue_observer.wait_for_queue_empty(
+                seconds=seconds,
+                processor_stop_signal_queue=processor_stop_signal_queue,
+            )
+            self._consumer.interrupt()
+        sec_left = None if seconds is None else seconds - (time() - ts)
+        self._consumer.join(sec_left)
+
+        # Close resources
+        self.close()
+
+        # Remove local files
+        if self._queue_observer.is_queue_empty():
+            self.cleanup()
+
+    def cleanup(self) -> None:
+        super().cleanup()
+        try:
+            self._processing_resources.data_path.rmdir()
```

Comment:
I think if you introduce new abstraction like `ProcessingResources` which itself is `WithResource` instance it makes more sense to delegate cleaning of path to `ProcessingResources` itself.

So just make `ProcessingResources` implement `cleanup` method in which it removes data path. And in `cleanup` of `AsyncOperationProcessor` just call `self._processing_resources.clenup()`

Also if you defined `processing_resources` method maybe its good idea to use it here.

(sorry for mansplaining lol)

Reply:
Good point